### PR TITLE
Enhanced filtration/sorting for tasks/projects/tasks in projects/cloud storages

### PR DESCRIPTION
**Pull request by bsekachev**
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

TODO:
- [x] Check import task/project features works as expected
- [x] Fix sorting resets after changing page 
- [x] Fix several resource calls when open a page
- [x] Fix filters for tasks in projects, they don't work [server side issue]

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: cvat-ui/src/actions/cloud-storage-actions.ts [hunk 1]
```diff
@@ -103,40 +103,16 @@ export type CloudStorageActions = ActionUnion<typeof cloudStoragesActions>;
...
-
         dispatch(cloudStoragesActions.getCloudStorages());
         dispatch(cloudStoragesActions.updateCloudStoragesGettingQuery(query));
 
         const filteredQuery = { ...query };
         for (const key in filteredQuery) {
-            if (filteredQuery[key] === null) {
-                delete filteredQuery[key];
+            if ((filteredQuery as { [index: string]: any })[key] === null) {
+                delete (filteredQuery as { [index: string]: any })[key];
```
Comment (klakhov) [inline cvat-ui/src/actions/cloud-storage-actions.ts:112]: Maybe declare interface for `{ [index: string]: any }`? I see there are the repetitions of this casting in other files

**Review by klakhov — COMMENTED** [review]
Hey, It's great to see new filtration/sorting panel
I've found several UX issues with filters:

- When two rules are present "or" doesn't have left border
image
- If we have group of rules and ungrouped rule, the line to the left is not straight. Also trash bin in group is a bit too high.
image
- Nested groups look too complicated for me, padding on the left seems insufficient to figure out what level the rule is  when there are three+ levels. Do we really need such depth?
image

File: cvat-ui/src/actions/cloud-storage-actions.ts (hunk 1 above)
Comment (bsekachev) [reply cvat-ui/src/actions/cloud-storage-actions.ts:112]: Makes sense

**Comment by ActiveChooN** [comment]
Great job! Very powerful solution.
Have a few UX comments/questions:

- It can be helpful to add the ability to create links for search/filter query for sharing between team members like frame URL
- It would be nice to see outside click filters dropdowns hiding
- Search query is saved between navigation, but filters are not

**Comment by nmanovic** [comment]
@bsekachev , some problems with Models page:

image

**Comment by nmanovic** [comment]
@bsekachev ,

> - It can be helpful to add the ability to create links for search/filter query for sharing between team members like frame URL, or maybe forward filter params to search query as it did in the GitHub search

It will be great if we can see applied filters in URL like https://github.com/openvinotoolkit/cvat/issues?q=is%3Aopen+is%3Aissue+label%3Abug

> - It would be nice to see outside click filters dropdowns hiding

Agree

> - Search query is saved between navigation, but filters are not

If we use correct URLs, the problem will be solved.

File: cvat-ui/src/components/resource-sorting-filtering/index.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import './styles.scss';
+import SortingComponent from './sorting';
+import ResourceFilterHOC from './filtering';
+
+const defaultVisibility: {
+    predefined: boolean;
```
Comment (ActiveChooN) [inline cvat-ui/src/components/resource-sorting-filtering/index.ts:10]: Probably type definition is redundant there

**Comment by ActiveChooN** [comment]
@nmanovic 
> It will be great if we can see applied filters in URL like

Probably this can be done as it is done in creating links to exact frame. Generating link with button and not to store this huge link in the search bar. And when users go via that link, redirect to `/<resource>` with corresponding filters.

File: cvat-ui/src/reducers/interfaces.ts [hunk 3]
```diff
@@ -45,7 +42,7 @@ export interface ProjectsState {
     count: number;
     current: Project[];
     gettingQuery: ProjectsQuery;
-    tasksGettingQuery: TasksQuery;
+    tasksGettingQuery: TasksQuery & { ordering: string };
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/reducers/interfaces.ts:45]: Relevant only for getting query in projects state

File: cvat-ui/src/actions/cloud-storage-actions.ts (hunk 1 above)
Comment (bsekachev) [reply cvat-ui/src/actions/cloud-storage-actions.ts]: Fixed

File: cvat-core/src/server-proxy.js [hunk 4]
```diff
@@ -425,6 +425,23 @@
                 return response.data.results;
             }
 
+            async function getProjectTasks(id, filter = {}) {
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-core/src/server-proxy.js]: No, removed.

File: cvat-ui/src/components/cloud-storages-page/top-bar.tsx [hunk 5]
```diff
@@ -1,42 +1,92 @@
...
     query: CloudStoragesQuery;
 }
 
 export default function StoragesTopBar(props: Props): JSX.Element {
-    const { onSearch, query } = props;
+    const {
+        query, onApplyFilter, onApplySorting, onApplySearch,
+    } = props;
     const history = useHistory();
+    const [visibility, setVisibility] = useState<typeof defaultVisibility>(defaultVisibility);
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/cloud-storages-page/top-bar.tsx]: Removed.

File: cvat-ui/src/components/cloud-storages-page/top-bar.tsx [hunk 6]
```diff
@@ -1,42 +1,92 @@
...
+            <Col span={24}>
+                <div className='cvat-cloudstorages-page-filters-wrapper'>
+                    <Input.Search
+                        enterButton
+                        onSearch={(phrase: string) => {
+                            onApplySearch(phrase);
+                        }}
+                        defaultValue={query.search || ''}
+                        className='cvat-cloudstorages-page-tasks-search-bar'
+                        placeholder='Search ..'
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/cloud-storages-page/top-bar.tsx]: Fixed

File: cvat-ui/src/components/projects-page/top-bar.tsx [hunk 7]
```diff
@@ -2,78 +2,135 @@
...
+    onApplySearch(search: string | null): void;
+    query: ProjectsQuery;
+    importing: boolean;
+}
+
+function TopBarComponent(props: Props): JSX.Element {
+    const {
+        importing, query, onApplyFilter, onApplySorting, onApplySearch, onImportProject,
+    } = props;
+    const [visibility, setVisibility] = useState<typeof defaultVisibility>(defaultVisibility);
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/projects-page/top-bar.tsx]: Removed.

File: cvat-ui/src/components/projects-page/top-bar.tsx [hunk 8]
```diff
@@ -2,78 +2,135 @@
...
+    const prevImporting = usePrevious<boolean>(importing);
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/projects-page/top-bar.tsx]: Removed.

File: cvat-ui/src/components/resource-sorting-filtering/index.ts (hunk 2 above)
Comment (bsekachev) [reply cvat-ui/src/components/resource-sorting-filtering/index.ts]: Removed.

**Review by ActiveChooN — DISMISSED** [review]

**Comment by nmanovic** [comment]
@bsekachev , can we show in browser top bar a readable URL like we do in recent filters? It is only for UI, thus we can have here more readable variant for our users. What do you think?

http://localhost:8080/jobs?filter=%7B%22%21%22%3A%7B%22or%22%3A%5B%7B%22%3D%3D%22%3A%5B%7B%22var%22%3A%22state%22%7D%2C%22completed%22%5D%7D%2C%7B%22%3D%3D%22%3A%5B%7B%22var%22%3A%22stage%22%7D%2C%22acceptance%22%5D%7D%5D%7D%7D&page=1

**Comment by nmanovic** [comment]
@bsekachev , if I clear all quick filters and refresh the project/task/job page, my decision isn't saved.

**Comment by bsekachev** [comment]
@nmanovic 

> can we show in browser top bar a readable URL like we do in recent filters? It is only for UI, thus we can have here more readable variant for our users. What do you think?

I do not actually think that somebody is going to change this filter via URL. This URL is intended to create a link to a specific request. 
What you suggest I think possible, but techically it brings some difficulties, for example we can't decode query string to filter internal representation, 3rdparty parser is necessary and maybe there are other unpredictable issues.

**Comment by bsekachev** [comment]
@nmanovic 

Now URLs are more readable:

image

They can be copied to clipboard and reopened with keeping filter. 

About default filters I would suggest followings:
- Do not apply default filters when open resources page, literally get rid of default filters 
- Remember the latest applied filter in local storage
- Filter from URL has higher priority than filter from local storage, so if we have two sources when open page, URL filter is restored

**Comment by nmanovic** [comment]
> @nmanovic
> 
> Now URLs are more readable:
> 
> image
> 
> They can be copied to clipboard and reopened with keeping filter.
> 
> About default filters I would suggest followings:
> 
> * Do not apply default filters when open resources page, literally get rid of default filters
> * Remember the latest applied filter in local storage
> * Filter from URL has higher priority than filter from local storage, so if we have two sources when open page, URL filter is restored

It looks reasonable. I agree with all proposals. The only question I have in my mind, can we unify the format of filters in URL and `recent`?

**Comment by bsekachev** [comment]
@nmanovic 

I implemented proposals, but it looks like making format in URL the same causes issues, related with restoring filter from such query strings.

We copy it, paste in another tab and there are not guarantees that we can restore filter as it was (because there are more types of operators in json format, and as I said before, ``react-awesome-query-builder`` does not provide solution to parse query strings). 

In theory we could look into 3rdparty parsers, but again, it does not guarantees as we do not meet issues in the future.

**Comment by klakhov** [comment]
@bsekachev 
Anyway, what do you think about changing delete button a bit (remove box-shadow and margin-top)?
pQN-L9P3QnQ

**Comment by bsekachev** [comment]
@klakhov 

These styles (margin, spaces, etc) are predefined by the library we are using. Maybe it is a good idea to create issue in their repository if you believe the interface looks confusing?

**Comment by nmanovic** [comment]
@bsekachev ,

Problem: a filter is active by it is not applied
- Build a filter on Jobs page
- Go to Tasks page
- Go to Jobs page

image

Possible solution: with the latest implementation it is better to reset filter when you leave a page (see how github issues filter works)

**Comment by bsekachev** [comment]
@nmanovic 

I discussed this issue with Andrey Chernov. He said it is not critical to keep the latest filter if there is an option to setup it in a couple of clicks (quick filters).

**Comment by bsekachev** [comment]
@nmanovic @ActiveChooN 

BTW, do you think it is important to keep sorting when go away from a page and then return back?
If we do not keep filters, it looks a bit inconsistent and GitHub also do not keep sorting in this case. 

Sorting and filters are kept if we just update page with F5 for example, because in this case they are read from URL.

**Review by nmanovic — APPROVED** [review]
LGTM

**Comment by bsekachev** [comment]
@TOsmanov Could you please update documentation for all the pages? (Jobs, Tasks, Projects, Tasks in a project, Cloud storages)

**Comment by TOsmanov** [comment]
> @TOsmanov Could you please update documentation for all the pages? (Jobs, Tasks, Projects, Tasks in a project, Cloud storages)

@bsekachev , Yes, I will update the documentation.

**Pull request merged**